[BD-24] [BB-2726] [TNL-7330] Added batch_get_or_create class method for ExternalId

shimulch opened the pull request:
edx/xblock-lti-consumer#124 depends on this PR.

This PR adds a new class method ``batch_get_or_create_user_ids`` to ``ExternalId``. This enables a faster way to create ``ExternalId`` for a list of ``User``.

When ``batch_get_or_create_user_ids`` gets called with a list of ``User`` instances and a ``type_name`` (Name of the type of ExternalId) it will return a dict mapping user id with corresponding external id.
```python
{
   user_id: ExternalId
}
```

**JIRA tickets**: This PR completes a part of the functionality required in the following tickets -
- https://openedx.atlassian.net/browse/TNL-7330
- https://tasks.opencraft.com/browse/BB-2726

**Dependencies**: None

**Merge deadline**: None

**Testing instructions**:

1. Pull this PR on your local devstack.
2. Run your local devstack
3. Open LMS shell by running ``make lms-shell``
4. Open Django shell for LMS ``python manage.py lms shell``
4. Create couple of users -
```python
from django.contrib.auth.models import User
users = [User.objects.create_user(username=f'test{i}', email=f'test{i}@beatles.com', password='test') for i in range(5)]
```
6. Import ``ExternalId`` model
```python
from openedx.core.djangoapps.external_user_ids.models import ExternalId
```
7. Create external ids in batch
```python
result = ExternalId.batch_get_or_create_user_ids(users, 'lti')
```
8. Check the result is in the above mentioned format. For example -
```python
{10: <ExternalId: ExternalId object (None)>, 11: <ExternalId: ExternalId object (None)>, 12: <ExternalId: ExternalId object (None)>, 13: <ExternalId: ExternalId object (None)>, 14: <ExternalId: ExternalId object (None)>}
```

**Author notes and concerns**:
N/A

**Reviewers**
- [ ] @giovannicimolin 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @shimulch! I've created [OSPR-5298](https://openedx.atlassian.net/browse/OSPR-5298) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

giovannicimolin commented on openedx/core/djangoapps/external_user_ids/models.py:
```diff
@@ -103,3 +103,38 @@ def add_new_user_id(cls, user, type_name):
+            )
+            return None
+
+        # get external ids that already exists
+        result = list(cls.objects.filter(user__in=users, external_id_type=type_obj))
+
+        # find users for those external ids needs to be created
+        users_wo_externalid = set(users) - {eid.user for eid in result}
```
I think this is not the optimal solution, given that a potentially big user dataset might come in.
Forcing a queryset evaluation and then subtracting two user objects sets might be an expensive operation.

shimulch replied:
@giovannicimolin, I've optimized it to use a query instead of set operation.

giovannicimolin reviewed (changes requested):
Left a few comments.

shimulch commented on openedx/core/djangoapps/external_user_ids/models.py:
```diff
@@ -103,3 +103,38 @@ def add_new_user_id(cls, user, type_name):
         return external_id, created
+
+    @classmethod
+    def batch_get_or_create(cls, users, type_name):
```
Updated.

shimulch commented on openedx/core/djangoapps/external_user_ids/models.py:
```diff
@@ -103,3 +103,38 @@ def add_new_user_id(cls, user, type_name):
+            ])
+            result += list(new_externalids)
+
+        return result
```
@giovannicimolin, Now the method returns a dictionary as per the suggested structure.

giovannicimolin replied:
@shimulch I think we can get rid of the created argument here. A simple `"user_id": ExternalId` map is enough here.

natabene commented:
@shimulch Thank you for your contribution. Please ping me once it is ready for edX review.

giovannicimolin commented on openedx/core/djangoapps/external_user_ids/tests/test_batch_generate_id.py:
```diff
@@ -0,0 +1,64 @@
+from openedx.core.djangoapps.external_user_ids.tests.factories import ExternalIDTypeFactory
+
+
+class TestBatchGenerateExternalIds(TestCase):
```
Since this is a batch operation, we should ideally add some query count test here, to make sure `more users != more queries`.

This can be done with a simple assertion in the `test_batch_get_or_create_user_ids` (see [`assertNumQueries`](https://docs.djangoproject.com/en/dev/topics/testing/tools/#django.test.TransactionTestCase.assertNumQueries), might require a change to `TransactionTestCase` instead of `TestCase`).

shimulch replied:
Tests with the expected number of queries added. Thanks for the tip 👍🏽

giovannicimolin reviewed (changes requested):
Aside a minor change in the data returned and a comment about a test, this is good to go.

bradenmacdonald commented on openedx/core/djangoapps/external_user_ids/models.py:
```diff
@@ -103,3 +103,50 @@ def add_new_user_id(cls, user, type_name):
+    @classmethod
+    def batch_get_or_create_user_ids(cls, users, type_name):
+        """
+        Create ExternalIds in batch.
```
Can you please provide an example of when it's useful to create these in a batch, instead of creating them for each user as needed?

shimulch replied:
@bradenmacdonald, In the case of LTI Consumer, External ID's are created when a student actually launches LTI from LMS. But when providing course member information to a third party tool, not every member has External ID's available. To create one by one would be a performance issue. Thus this method has been created.

bradenmacdonald replied:
@shimulch Oops, I meant to specify that I'd like that added to the docstring. Dosctrings should generally say both _what_ the code does and _why_ it exists, because without the why it's hard to know when/how it can be refactored or removed in the future.

natabene commented:
@shimulch Just checking if you are still planning on pursuing this PR.

pomegranited commented:
@natabene CC @bradenmacdonald @shimulch  This PR is related to https://github.com/edx/xblock-lti-consumer/pull/124, which is for a Blended project. Could someone tag it so we know it's not an ordinary OSPR?

giovannicimolin commented:
@natabene As @pomegranited mentioned, this is related to BD-24. We ended up having discussions related to the implementation on [this ADR](https://github.com/edx/xblock-lti-consumer/pull/142) before continuing the work. We're now unblocked and will start moving this forward in the coming week.

giovannicimolin reviewed (changes requested):
@shimulch Code is working perfectly, but I have a few nits and changes:
1. Python APIs should not return internal classes whenever possible. Please change the `ExternalId` object on the response with the plain text external id. I'm sorry that wasn't clear in the comment.  
1. Improve this PR test instructions and proper testing instructions for verifying this works on the devstack.
2. Clean up the unused description fields.
3. Rebase this PR on the latest `master`.

shimulch commented on openedx/core/djangoapps/external_user_ids/models.py:
```diff
@@ -103,3 +103,59 @@ def add_new_user_id(cls, user, type_name):
+        LTI from LMS. But when providing course member information
+        to a third party tool, not every member has External ID's
+        available. To create one by one would be a performance issue.
+        This method provides a faster way to create ExternalIds in batch.
```
Removed, LTI specific doc, and added generic doc which explains what this method does.

shimulch commented:
@giovannicimolin,

> Python APIs should not return internal classes/objects whenever possible. Please change the ExternalId object on the response with the plain text external id.

This method is just a batch version of the existing `add_new_user_id` method which returns the `ExternalId` object as well.

> Improve this PR test instructions and proper testing instructions for verifying this works on the devstack.

Thanks for pointing that out. I've added proper test instruction on how to verify the new method.

> Clean up the unused description fields.

If you are talking about the PR description, then yes, I've cleaned up. If that's not the case I am probably not sure what you meant.

> Rebase this PR on the latest master.

Rebased.

giovannicimolin reviewed (approved):
:+1:

- [x] I tested this: 
1. Followed the testing instructions on the devstack.
2. Checked that the query count didn't increase when increasing the number of users created (tested with 2, 20, and 200 users in the test script above).
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ NA
- [x] Includes documentation: In docstrings

@shimulch This is good to go. Great work.
@nedbat This is ready for your review.

nedbat commented on openedx/core/djangoapps/external_user_ids/models.py:
```diff
@@ -103,3 +103,58 @@ def add_new_user_id(cls, user, type_name):
+            type_obj = ExternalIdType.objects.get(name=type_name)
+        except ExternalIdType.DoesNotExist:
+            LOGGER.info(
+                'Batch ID Creation failed, no external id type of {type}'.format(
```
```suggestion
                'Batch ID Creation failed, no external id type of {type!r}'.format(
```

shimulch replied:
Updated.

nedbat commented on openedx/core/djangoapps/external_user_ids/models.py:
```diff
@@ -103,3 +103,58 @@ def add_new_user_id(cls, user, type_name):
+                    type=type_name
+                )
+            )
+            return None
```
Should this be `{}` instead to stick to the signature of the function?

nedbat replied:
Or would it be better to raise an exception? I don't know about ExternalIdType, and what callers expect.

nedbat replied:
(or at least update the docstring)

shimulch replied:
The reasoning behind sending `None` is to keep it similar to `add_new_user_id` which also returns `None` if fails. I've updated the docstring.

nedbat commented on openedx/core/djangoapps/external_user_ids/tests/test_batch_generate_id.py:
```diff
@@ -0,0 +1,67 @@
+        with self.assertNumQueries(self.EXPECTED_NUM_OF_QUERIES):
+            result = ExternalId.batch_get_or_create_user_ids(users, id_type)
+
+        assert len(result) == len(result)
```
This surely is not what you meant!
```suggestion
        assert len(result) == len(users)
```

shimulch replied:
Sorry for that and thanks for catching it.

nedbat reviewed (changes requested):
GitHub says I have to leave a comment explaining the changes, so I will point you to all of the comments I left explaining changes :)

shimulch commented:
@nedbat Thanks for the review. I've addressed your comments. :-)
